crash-recovery/README.md

```diff
@@ -0,0 +1,78 @@
+Unexpect crashes are bound to occur when BigchainDB is deployed in production which could leave the system in an inconsistant state. Hence it is necessary that an appropriate methodology should be developed to restore the system to a consistant state.
+
+
+## Motivation
+For BigchainDB, a crash which can cause the state to be inconsistent could leave the system inoperable. The 'achilles heel' for BigchainDB is its `commit` function during which a new block with its transactions is written to the disk. Furthermore, UTXO merkle tree is updated based on these new transactions. An overview of what happens during `commit` is given below,
+
+1. Store metadata of the transactions.
+1. Store assets of the transactions.
+1. Store transactions and update UTXO.
+1. Store Block.
+
+These write operations are not transactional so any intermediate failure could result in an inconsistent state. For example, 
+
+- Consider a scenario where the system fails after step 2 i.e. we have a state where such metadata and assets exists which don't belong to any transaction (also called zombie assets and zombie metadata).
+- Consider a scenario wherein the system fails during setup 3 i.e. we would have a state which zombie metadata, zombie assets and zombie transactions (because the block to which these transactions belong is written in 4.). Moreover we might have an inconsistent UTXO.
+
+Hence it is crucial that the system should be able to recover from such crashes and bring itself to a consistent state.
+
+
+## Specification
+The ABCI proxy app adheres to the following state machine to commit a new block,
+
+```
+BEGIN BLOCK
+    ↓
+DELIVER TRANSACTION
+    ↓
+END BLOCK
+    ↓
+COMMIT
+```
+
+Here `BEGIN BLOCK` is used to indicate a new incoming block and `DEVLIER TRANSACTION` is used to iteratively deliver each transaction in this new block. `END BLOCK` is called by Tendermint when the set of transactions being commit-ed in the new block have been delivered (NOTE: the block height of the new block is passed as argument to `end_block`). Lastly, `COMMIT` is called and it is expected that all the state mutations are done during `COMMIT`. So, as previously discussed the transactions are stored and the block is added in the database.
+
+Since `COMMIT` api is mission critical and any crash during `COMMIT` can cause the system to be inconsistent so it is necessary that there should be a pre-commit state stored to the disk which should store the metadata of the state mutations being performed during actual `COMMIT` i.e. pre-commit object should at least store the following data,
+
+```json
+{ "id": "pre_commit",
+  "height": 10,
+  "transaction_ids": ["txid1", "txid2", "txidn"],
+}
+```
+
+The above object should be stored in a separate collection `pre_commit` during `END BLOCK`. Note that the `pre_commit` collection should have only 1 object with `'id': 'pre_commit'` because next write to `pre_commit` collection implies that the previous block was successfully `commit`ed. 
```

Comment:
What is the initial value of `pre_commit`?

Reply:
@vrde That is a good question. As of now it is `None`